### PR TITLE
Register SyncBatchNorm as quantization module

### PR DESCRIPTION
**Pull request by 5had3z**
### What does this PR do?

Type of change: Bug fix 

Registers `nn.SyncBatchNorm` layer for quantization. If a model is setup for distributed training before PTQ, none of the SyncBatchNorm layers are recognised and quantized. On loading of a checkpoint there is now a mismatch between the modelopt state of a model that hasn't had DDP/SyncBN applied to it and the checkpoint trained with DDP/SyncBN.

Performing PTQ and then applying DDP/SyncBN for QAT works fine, but considering that unwrapping DDP is handled properly for either ordering of the steps, SyncBN conversion should be able to be performed in either order as well.

### Usage

```python
## train.py
model = get_model()
# DDP Setup
nn.SyncBatchNorm.convert_sync_batchnorm(model)
model = nn.parallel.DistributedDataParallel(
    model, device_ids=[dist.get_rank()], output_device=dist.get_rank()
)
# PTQ
mtq.quantize(model, mtq.INT8_DEFAULT_CFG, calib)
mtq.print_quant_summary(model) # Missing SyncBN layers
# QAT
train(model)
# Save Checkpoint
torch.save(mto.modelopt_state(model), "modelopt.pt")
torch.save(model.module.state_dict(), "params.pt")

## inference.py
model = get_model()
# Below fails as nn.BatchNorm2d in current model state does not have state in checkpoint since 
# nn.SyncBatchNorm modules were skipped over.
model = restore_model_from_modelopt_state(model, torch.load("modelopt.pt", weights_only=False))
mode.load_state_dict(torch.load("params.pt")
```

### Testing

Added `nn.SyncBatchNorm` to the quantization tests where other BatchNorm layers appear..

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?:  ❌ - Models trained with missing norm from their modelopt state dict will now have this depending on initialization order. 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: ✅
- Did you write any new necessary tests?: ✅
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ❌ - lmk if you want this
- Did you get Claude approval on this PR?:  ❌ 
 
### Additional Information

Code for testing issue, run with `python3 script.py` or `torchrun --nproc-per-node=2 script.py`.

```python
from pathlib import Path
import torch
import os
from torch import nn
import torch.distributed as dist
from torchvision.models import resnet18, ResNet18_Weights
from torchvision.datasets.cifar import CIFAR10
from torch.utils.data import DataLoader, DistributedSampler
from torchvision.transforms.v2 import ToTensor
import modelopt.torch.quantization as mtq
import modelopt.torch.opt as mto
from rich.progress import track

assert torch.cuda.is_available(), "NVIDIA GPUs required for distributed training"

torch.cuda.set_device(f"cuda:{os.environ.get('LOCAL_RANK', 0)}")
if dist.is_available() and int(os.environ.get("WORLD_SIZE", 1)) > 1:
    dist.init_process_group(backend="nccl")

model = resnet18(weights=ResNet18_Weights.DEFAULT).cuda()

if dist.is_initialized():  # SyncBN and DDP for training
    nn.SyncBatchNorm.convert_sync_batchnorm(model)
    model = nn.parallel.DistributedDataParallel(
        model, device_ids=[dist.get_rank()], output_device=dist.get_rank()
    )


def calib(m: nn.Module):
    datapath = Path.cwd() / "data"
    datapath.mkdir(exist_ok=True)
    dataset = CIFAR10(datapath, train=False, download=True, transform=ToTensor())
    if dist.is_initialized():
        sampler = DistributedSampler(dataset)
    else:
        sampler = None
    dataloader = DataLoader(dataset, sampler=sampler, num_workers=4, batch_size=8)
    for img, tgt in track(dataloader):
        m(img.cuda())


# PTQ
mtq.quantize(model, mtq.INT8_DEFAULT_CFG, calib)
if not dist.is_initialized() or dist.get_rank() == 0:
    mtq.print_quant_summary(model)

# Do training

mto_ckpt = Path.cwd() / "opt.pt"
torch.save(mto.modelopt_state(model), mto_ckpt)
param_ckpt = Path.cwd() / "params.pt"
if isinstance(model, nn.parallel.DistributedDataParallel):
    params = model.module.state_dict()
else:
    params = model.state_dict()
torch.save(params, param_ckpt)

# Load 'single' GPU for inference
model = resnet18(weights=ResNet18_Weights.DEFAULT).cuda()
model = mto.restore_from_modelopt_state(
    model, torch.load(mto_ckpt, map_location="cuda", weights_only=False)
)
model.load_state_dict(torch.load(param_ckpt))

```

## Summary by CodeRabbit

* **New Features**
  * Added support for synchronized batch normalization layers in quantization workflows.

* **Tests**
  * Extended quantization test coverage for synchronized batch normalization module types.

* **Chores**
  * Updated quantization configurations to handle synchronized batch normalization layers consistently with other batch norm types.

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA/Model-Optimizer/pull/1491)

### CODE REVIEW

**Comment by meenchen** [comment]
@5had3z thanks for the PR. Could you share some background on the use cases for BatchNorm quantization?

**Comment by 5had3z** [comment]
@meenchen `SyncBatchNorm` is disabled by default, just like the other batch norms. The main purpose of this change is to resolve the problem is illustrated in the example code where if `nn.SyncBatchNorm.convert_sync_batchnorm` is applied to the model for DDP training before the quantization conversion, none of the normalization layers are registered. When resuming for evaluation on a single GPU, `restore_from_modelopt_state` will register the conversion of the normalization layers (still disabled by default). Then when it tries to copy the saved modelopt state dict to the model, there is a mismatch in the current state dict (containing norm layers) and the saved state dict (missing norm layers) and an error is raised in `torch.quantization.restore_quantizer_state` where `extra_keys` is not empty.

EDIT: Sorry, I thought I had comments of where the problems occur in the example code. Maybe I tweaked something in the test script, re-copied it over and was missing the comments.

EDIT2: Ahh no yes I have comments in the first block, just not the second that is the full end-to-end test code.

**Review by cjluo-nv — APPROVED** [review]
> _Bot review — DM the bot to share feedback._

Small, well-scoped bug fix that registers `nn.SyncBatchNorm` analogously to the existing `BatchNorm1d/2d/3d` registrations. The two YAML configs that enumerate BatchNorm variants are both updated consistently, and the existing parameterized batchnorm tests are extended to cover SyncBatchNorm across `test_no_quant`, `test_fake_quant_per_tensor`, and `test_fake_quant_per_channel`. Explicit registration is necessary since `QuantModuleRegistry` uses exact class matching. The PR body documents a clear repro and notes the back-compat implication for checkpoints saved before this fix.

**Review by meenchen — COMMENTED** [review]
### Correctness — Incomplete Coverage of Exclusion Lists

**[BLOCKER] SyncBatchNorm not added to `_default_disabled_quantizer_cfg` in `config.py`**
The hardcoded list at `modelopt/torch/quantization/config.py:211-213` excludes `nn.BatchNorm{1,2,3}d` but not `nn.SyncBatchNorm`. This list is used by `FP8_DEFAULT_CFG`, `INT8_DEFAULT_CFG`, and other built-in `*_CFG` dicts. After this PR, any user using these configs with a SyncBN-containing model will silently have SyncBN layers quantized — changing behavior for non-DDP users too.

**[BLOCKER] 5 general PTQ recipes still have inline BatchNorm exclusions without SyncBatchNorm**
The PR updates the `$import` unit (`default_disabled_quantizers.yaml`) and one model-specific recipe (`Step3.5-Flash`). But these recipes have **inline** exclusion lists that don't yet use the unit:
- `modelopt_recipes/general/ptq/fp8_default-fp8_kv.yml`
- `modelopt_recipes/general/ptq/nvfp4_default-fp8_kv.yml`
- `modelopt_recipes/general/ptq/nvfp4_experts_only-fp8_kv.yml`
- `modelopt_recipes/general/ptq/nvfp4_mlp_only-fp8_kv.yml`
- `modelopt_recipes/general/ptq/nvfp4_omlp_only-fp8_kv.yml`

Each needs a `- parent_class: 'nn.SyncBatchNorm'` entry. Without these updates, SyncBN layers will be quantized when these recipes are used — inconsistent behavior depending on recipe choice.

**Comment by 5had3z** [comment]
> **[BLOCKER] 5 general PTQ recipes still have inline BatchNorm exclusions without SyncBatchNorm**

All of these configs should have imported the changes I have already made in `modelopt_recipes/configs/ptq/units/default_disabled_quantizers.yaml`. I'm not sure what else needs to change.

```yaml
imports:
    default_disabled_quantizers: configs/ptq/units/default_disabled_quantizers
```

**Comment by 5had3z** [comment]
@meenchen SyncBatchNorm is registered as a default disabled module. I pip installed my local repo and can see in  `modelopt_recipes.configs.ptq.units` that syncbatchnorm has been added to `default_distabled_quantizers.yaml`. It has also been added, ad hoc similar to the other batch norm layers, in `int4_blockwise_weight_only.yaml` and `nvfp4-mlp-only.yaml`. Where else should it be appearing?
